File: 4.8-bc/ts/stage-2-feature-first/src/features/auth/service.ts
// Auth feature — service: реєстрація, логін.
// Vertical slice: handler, service, repository, model — всі в одній папці.
import { randomUUID } from "node:crypto";
import bcrypt from "bcryptjs";

import type { User } from "./model.js";
import type { Repository } from "./repository.js";

export class InvalidCredentialsError extends Error {
  constructor() {
    super("auth.invalid_credentials");
    this.name = "InvalidCredentialsError";
  }
}

export class Service {
  constructor(private readonly repo: Repository) {}

  async register(email: string, password: string): Promise<User> {
    const passwordHash = await bcrypt.hash(password, 10);
    const u: User = {
      id: randomUUID(),
      email,
      passwordHash,
      createdAt: new Date(),
    };
    await this.repo.create(u);
    return u;
  }

  async login(email: string, password: string): Promise<User> {
    const u = await this.repo.findByEmail(email);
    if (!u) throw new InvalidCredentialsError();
    const ok = await bcrypt.compare(password, u.passwordHash);
    if (!ok) throw new InvalidCredentialsError();
    return u;
  }
}

File: 4.8-bc/py/stage-2-feature-first/app/features/auth/service.py
"""Auth feature — registration, login. Vertical slice service.

Sentinel errors map to ``AppError`` in handler.py.
"""
from __future__ import annotations

from datetime import UTC, datetime
from uuid import uuid4

from passlib.hash import bcrypt

from app.features.auth.model import User
from app.features.auth.repository import Repository


class InvalidCredentialsError(Exception):
    """auth.invalid_credentials"""


class Service:
    def __init__(self, repo: Repository) -> None:
        self._repo = repo

    async def register(self, email: str, password: str) -> User:
        password_hash = bcrypt.hash(password)
        u = User(
            id=uuid4(),
            email=email,
            password_hash=password_hash,
            created_at=datetime.now(UTC),
        )
        await self._repo.create(u)
        return u

    async def login(self, email: str, password: str) -> User:
        u = await self._repo.find_by_email(email)
        if u is None:
            raise InvalidCredentialsError()
        if not bcrypt.verify(password, u.password_hash):
            raise InvalidCredentialsError()
        return u
